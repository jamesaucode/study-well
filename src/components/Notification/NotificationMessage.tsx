import React from "react";
import styled from "styled-components";
import { Message } from "../../styles/shared";
import { NextFC } from "next";
import { IMessage } from "../../../interfaces/message";
import SVG from "react-inlinesvg";

interface WrapperProps {
  mounted?: boolean;
}
const Wrapper = styled.div<WrapperProps>`
  display: flex;
  justify-content: center;
  border-radius: 5px;
  width: 100%;
  padding: 0.75rem;
  box-sizing: border-box;
  transition: 0.5s ease-in-out opacity;
`;
const Important = styled.span`
  font-weight: 700;
`;
const SVGWrapper = styled.div`
  padding: 0 0.5rem;
`;
const renderImportant = (text: string): React.ReactElement => (
  <Important>{text}</Important>
);
const NotificationMessage: NextFC<IMessage> = ({
  success,
  message,
  id,
  // delay = 3000,
  removeNotification,
}): JSX.Element => {
  return (
    <Wrapper key={id}>
      <Message success={success}>
        <div>
          {success ? renderImportant("Success! ") : renderImportant("Oops! ")}
          {message}
        </div>
        <SVGWrapper onClick={(): void => removeNotification(id)}>
          <SVG className="small-icon" src="/static/images/close.svg" />
        </SVGWrapper>
      </Message>
    </Wrapper>
  );
};

NotificationMessage.defaultProps = {
  success: false,
  message: "No message",
};

export default React.memo(NotificationMessage);
