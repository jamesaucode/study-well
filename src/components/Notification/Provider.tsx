import React, { useRef, useState, useEffect } from 'react';
import { Context } from './Context';

const Provider = () => {
    const root  = useRef<Node | null>(null);
    const id = useRef([]);
    const [notifications, setNotifications] = useState([]);

    useEffect(() => {
        root.current = document.createElement('div');
        document.body.appendChild(root.current);
        return () => {
           id.current.forEach(clearTimeout);
           if (root.current) {
            document.body.removeChild(root.current);
           }
        };
    }, [])

    // const remove = (notification) => {

    // }
}
