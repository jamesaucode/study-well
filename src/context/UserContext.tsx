import React, { createContext, useContext } from "react";
import styled from "styled-components";
import useFetchData from "../hooks/useFetchData";
import { UserTypes } from "../../resources/user/user.model";
import { HeadingBase } from "../../utils/style";
import Loading from "../components/Loading";

const Wrapper = styled.div`
  height: 100vh;
  width: 100vw;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`;

export const UserContext = createContext<UserTypes | {} | null>(null);

export const UserProvider: React.FunctionComponent = ({
  children,
}): JSX.Element => {
  const [{ data, isLoading, isError }] = useFetchData("/api/user");
  return isLoading ? (
    <Wrapper>
      <HeadingBase>Loading</HeadingBase>
      <Loading />
    </Wrapper>
  ) : (
    <UserContext.Provider value={data}>
      {isError && (
        <HeadingBase>Something went wrong during data fetching...</HeadingBase>
      )}
      {children}
    </UserContext.Provider>
  );
};

export const useUserContext = (): UserTypes | {} => useContext(UserContext);
