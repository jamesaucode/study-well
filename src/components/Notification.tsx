import React, { useState, useEffect, useRef } from "react";
import NotificationMessage from "./NotificationMessage";
import { MessageType } from "../../typings/message";

const Notification = (props: any) => {
    const [notifications, setNotifications] = useState();
    const popNotification = () => {
        setNotifications(notifications.slice(0, -1));
    }
    const pushNotification = (message  = "lmaooooo", status: boolean) => {
        setNotifications(notifications.concat({
            message,
            success: status
        }))
    }

  const children = React.Children.map(props.children, child => {
      return React.cloneElement(child, {
          popNotification,
          pushNotification 
      })
  })
  if (notifications) {
    return (
      <React.Fragment>
        {notifications.map((notification: MessageType) => {
          return (
            <NotificationMessage
              message={notification.message}
              success={notification.success}
            />
          );
        })}
        {children}
      </React.Fragment>
    );
  } else {
    return <button onClick={popNotification}>Test Button</button>;
  }
};

export default React.memo(Notification);
