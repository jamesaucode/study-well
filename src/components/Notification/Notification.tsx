import React, { useState, useEffect, useRef, EffectCallback } from "react";
import NotificationMessage from "./NotificationMessage";
import { IMessage } from "../../../interfaces/message";
import styled from "styled-components";
import { useWindowSize } from "../../hooks/useWindowSize";
import uuid from "uuid";

interface PropTypes {
  children: React.ReactChild;
}

interface AllNotificationWrapperProps {
  windowHeight: number;
  os: number;
}

interface NotificationContextProps {
  pushNotification: (message: string, success: boolean) => void;
  removeNotification: (message: string, success: boolean) => void;
}
const AllNotificationsWrapper = styled.div<AllNotificationWrapperProps>`
  height: 0;
  top: 0px;
  font-size: calc(0.35vw + 16px);
  width: 100%;
  position: absolute;
  z-index: 1000;
`;
const NotificationContext = React.createContext<NotificationContextProps | null>(
  null,
);
export const NotificationProvider: React.FunctionComponent<PropTypes> = ({
  children,
}): JSX.Element => {
  const [notifications, setNotifications] = useState<IMessage[]>([]);
  const windowSize = useWindowSize();
  const delay = 5000;
  const id = useRef<number[]>([]);
  useEffect((): EffectCallback => {
    return (): void => {
      id.current.forEach(clearTimeout);
    };
  }, []);
  const removeNotification = (notificationId: string): void => {
    setNotifications((prevState): IMessage[] => {
      return prevState.filter((n): boolean => n.id !== notificationId);
    });
  };
  const pushNotification = (message = "", status: boolean): IMessage => {
    const notification = {
      message,
      success: status,
      id: uuid(),
      delay,
    };
    const timerId = setTimeout((): void => {
      removeNotification(notification.id);
      id.current.splice(id.current.indexOf(timerId), 1);
    }, delay);
    id.current.push(timerId);
    setNotifications((allNotification): IMessage[] =>
      allNotification.concat(notification),
    );
    return notification;
  };

  const value = {
    pushNotification,
    removeNotification,
  };
  if (notifications) {
    return (
      <NotificationContext.Provider value={value}>
        <AllNotificationsWrapper
          windowHeight={windowSize.size.windowHeight}
          os={notifications.length * 54}
        >
          {notifications.map(
            (notification: IMessage): React.ReactElement => {
              return (
                <NotificationMessage
                  key={notification.id}
                  {...notification}
                  removeNotification={removeNotification}
                />
              );
            },
          )}
        </AllNotificationsWrapper>
        {children}
      </NotificationContext.Provider>
    );
  } else {
    return <React.Fragment>{children}</React.Fragment>;
  }
};

export const useNotification = (): NotificationContextProps =>
  React.useContext(NotificationContext);
