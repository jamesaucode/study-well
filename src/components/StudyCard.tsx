import React, { useState, useRef, useEffect } from "react";
import styled from "styled-components";
import { Button, font, colors } from "../../utils/style";
import { useNotification } from "../components/Notification/Notification";
import { useUserData } from "../hooks/useUserData";
import DropDownMenu from "./DropDownMenu";

const CardWrapper = styled.div`
  box-shadow: 0px 0px 20px ${colors.black}25;
  border-radius: 3px;
  margin-bottom: 20px;
  display: flex;
  flex-direction: column;
  background: #fff;
  font-size: calc(0.35vw + 16px);
  width: 100%;
  height: auto;
`;
const InputArea = styled.textarea`
  width: 100%;
  border: none;
  background: transparent;
  color: #333;
  flex: 1 auto;
  font-size: 1em;
  font-family: ${font.fontFamily};
  padding: 0.5em 1em;
  resize: none;
  box-sizing: border-box;
`;
const Wrapper = styled.section`
  width: 100%;
  max-width: 688px;
  padding: 1rem;
`;
const StudyCard: React.FunctionComponent = (): JSX.Element => {
  const { pushNotification } = useNotification();
  const { data } = useUserData();

  const [question, setQuestion] = useState("");
  const [answer, setAnswer] = useState("");
  const [selected, setSelected] = useState();
  const dropDownInputRef = useRef<HTMLInputElement>();

  const formControls: {
    [key: string]: React.Dispatch<React.SetStateAction<string>>;
  } = {
    question: setQuestion,
    answer: setAnswer,
    selected: setSelected,
  };
  const changeHandler = ({
    target: { name, value },
  }: React.ChangeEvent<HTMLTextAreaElement | HTMLSelectElement>): void => {
    formControls[name](value);
  };
  const clearInput = (): void => {
    setQuestion("");
    setAnswer("");
  };
  const handleSubmit = async (): Promise<void> => {
    try {
      event.preventDefault();
      const response = await fetch(`/api/cardset/${selected}`, {
        method: "POST",
        credentials: "include",
        headers: {
          "content-type": "application/json",
        },
        body: JSON.stringify({
          question,
          answer,
        }),
      });
      const json = await response.json();
      if (response.ok && json) {
        pushNotification(json.message, true);
        clearInput();
      }
    } catch (error) {
      pushNotification(error.message, false);
    }
  };
  const handleKeyDown = (
    event: React.KeyboardEvent<HTMLTextAreaElement>,
  ): void => {
    if (event.keyCode === 13) {
      handleSubmit();
      clearInput();
    }
  };
  const handleOptionClick = (
    event: React.MouseEvent<HTMLSpanElement>,
  ): void => {
    setSelected((event.target as HTMLElement).textContent);
  };
  const handleOptionSubmit = async (
    event: React.MouseEvent<HTMLButtonElement>,
  ): Promise<void> => {
    try {
      event.preventDefault();
      const response = await fetch("/api/cardset", {
        method: "POST",
        credentials: "include",
        headers: {
          "content-type": "application/json",
        },
        body: JSON.stringify({
          name: dropDownInputRef.current.value,
        }),
      });
      const json = await response.json();
      console.log(json);
    } catch (error) {
      console.error(error);
    }
  };

  const hasExistingCardsets = (data && data.cardSet.length > 0) || false;
  return (
    <Wrapper>
      <DropDownMenu
        initSelect={hasExistingCardsets ? data.cardSet[0].name : null}
      >
        <DropDownMenu.MenuTitle>Choose a set</DropDownMenu.MenuTitle>
        {hasExistingCardsets
          ? data.cardSet.map(
              (c): JSX.Element => (
                <DropDownMenu.Option
                  key={c._id}
                  extraOnClick={handleOptionClick}
                >
                  {c.name}
                </DropDownMenu.Option>
              ),
            )
          : null}
        <DropDownMenu.Input
          ref={dropDownInputRef}
          placeholder="Create a new set"
        />
        <DropDownMenu.MenuButton onClick={handleOptionSubmit}>
          Create
        </DropDownMenu.MenuButton>
      </DropDownMenu>
      <CardWrapper>
        <InputArea
          onChange={changeHandler}
          onKeyDown={handleKeyDown}
          value={question}
          name="question"
          placeholder="Question"
        />
        <InputArea
          onChange={changeHandler}
          onKeyDown={handleKeyDown}
          value={answer}
          name="answer"
          placeholder="Answer"
        />
      </CardWrapper>
      <Button fullWidth onClick={handleSubmit}>
        Save
      </Button>
    </Wrapper>
  );
};

export default StudyCard;
