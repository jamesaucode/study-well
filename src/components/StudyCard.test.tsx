/* eslint-disable */
import { mount } from "enzyme";
import StudyCard from "./StudyCard";
import { NotificationProvider } from "./Notification/Notification";
import { UserProvider } from "../context/UserContext";

describe("StudyCard", () => {
  // const mockPushNotification = jest.fn(() => {});
  it("Component renders", () => {
    const wrapper = mount(
      <UserProvider>
        <NotificationProvider>
          <StudyCard />
        </NotificationProvider>
      </UserProvider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
