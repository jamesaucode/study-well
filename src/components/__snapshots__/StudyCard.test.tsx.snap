// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudyCard Component renders 1`] = `
<UserProvider>
  <style__HeadingBase>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "style__HeadingBase-oz4l5n-3",
            "isStatic": false,
            "lastClassName": "cpnjeR",
            "rules": Array [
              "text-align:",
              [Function],
              ";font-size:",
              [Function],
              ";font-weight:600;margin:1rem 0;color:",
              [Function],
              ";",
            ],
          },
          "displayName": "style__HeadingBase",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "style__HeadingBase-oz4l5n-3",
          "target": "h1",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <h1
        className="style__HeadingBase-oz4l5n-3 cpnjeR"
      >
        Something went wrong during data fetching...
      </h1>
    </StyledComponent>
  </style__HeadingBase>
  <NotificationProvider>
    <Notification__AllNotificationsWrapper
      os={0}
      windowHeight={768}
    >
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "Notification__AllNotificationsWrapper-sc-14z8rhj-0",
              "isStatic": true,
              "lastClassName": "kdjGqQ",
              "rules": Array [
                "height:0;top:0px;font-size:calc(0.35vw + 16px);width:100%;position:absolute;z-index:1000;",
              ],
            },
            "displayName": "Notification__AllNotificationsWrapper",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "Notification__AllNotificationsWrapper-sc-14z8rhj-0",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
        os={0}
        windowHeight={768}
      >
        <div
          className="Notification__AllNotificationsWrapper-sc-14z8rhj-0 kdjGqQ"
        />
      </StyledComponent>
    </Notification__AllNotificationsWrapper>
    <StudyCard>
      <StudyCard__Wrapper>
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "StudyCard__Wrapper-sc-77egjm-2",
                "isStatic": true,
                "lastClassName": "kPwpOv",
                "rules": Array [
                  "width:100%;max-width:688px;padding:1rem;",
                ],
              },
              "displayName": "StudyCard__Wrapper",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "StudyCard__Wrapper-sc-77egjm-2",
              "target": "section",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <section
            className="StudyCard__Wrapper-sc-77egjm-2 kPwpOv"
          >
            <DropDownMenu
              initSelect={null}
            >
              <DropDownMenu__Wrapper
                role="menu"
              >
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "DropDownMenu__Wrapper-sc-1pyz6ng-1",
                        "isStatic": true,
                        "lastClassName": "irgNyQ",
                        "rules": Array [
                          "width:100%;max-width:250px;position:relative;",
                        ],
                      },
                      "displayName": "DropDownMenu__Wrapper",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "DropDownMenu__Wrapper-sc-1pyz6ng-1",
                      "target": "div",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  role="menu"
                >
                  <div
                    className="DropDownMenu__Wrapper-sc-1pyz6ng-1 irgNyQ"
                    role="menu"
                  >
                    <DropDownMenu__Toggle
                      onClick={[Function]}
                    >
                      <StyledComponent
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "DropDownMenu__Toggle-sc-1pyz6ng-3",
                              "isStatic": true,
                              "lastClassName": "dUSfAA",
                              "rules": Array [
                                "border:0;border-radius:3px;color:",
                                "#222222",
                                ";padding:10px 20px;margin-bottom:1rem;font-size:0.8em;font-weight:600;font-family:",
                                "-apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"",
                                ";text-align:left;z-index:100;&:hover{cursor:pointer;}",
                              ],
                            },
                            "displayName": "DropDownMenu__Toggle",
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "DropDownMenu__Toggle-sc-1pyz6ng-3",
                            "target": "button",
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                        onClick={[Function]}
                      >
                        <button
                          className="DropDownMenu__Toggle-sc-1pyz6ng-3 dUSfAA"
                          onClick={[Function]}
                        >
                          null â–¼
                        </button>
                      </StyledComponent>
                    </DropDownMenu__Toggle>
                    <DropDownMenu__MenuWrapper
                      expanded={false}
                    >
                      <StyledComponent
                        expanded={false}
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "DropDownMenu__MenuWrapper-sc-1pyz6ng-2",
                              "isStatic": false,
                              "lastClassName": "LmCiw",
                              "rules": Array [
                                "border-bottom-left-radius:4px;border-bottom-right-radius:4px;background-color:#fff;box-shadow:0px 0px 20px rgba(0,0,0,0.15);display:flex;flex-direction:column;padding:",
                                [Function],
                                ";position:absolute;top:40px;width:100%;z-index:100;transition:transform 300ms ease;transform-origin:top left;transform:",
                                [Function],
                                ";input{border:none;border-radius:3px;background-color:#eee;font-size:",
                                "0.8em",
                                ";padding:10px;margin:10px 0;&:hover{background-color:#ddd;}}",
                              ],
                            },
                            "displayName": "DropDownMenu__MenuWrapper",
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "DropDownMenu__MenuWrapper-sc-1pyz6ng-2",
                            "target": "ul",
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                      >
                        <ul
                          className="DropDownMenu__MenuWrapper-sc-1pyz6ng-2 LmCiw"
                        />
                      </StyledComponent>
                    </DropDownMenu__MenuWrapper>
                  </div>
                </StyledComponent>
              </DropDownMenu__Wrapper>
            </DropDownMenu>
            <StudyCard__CardWrapper>
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "StudyCard__CardWrapper-sc-77egjm-0",
                      "isStatic": true,
                      "lastClassName": "dzZgLX",
                      "rules": Array [
                        "box-shadow:0px 0px 20px ",
                        "#222222",
                        "25;border-radius:3px;margin-bottom:20px;display:flex;flex-direction:column;background:#fff;font-size:calc(0.35vw + 16px);width:100%;height:auto;",
                      ],
                    },
                    "displayName": "StudyCard__CardWrapper",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "StudyCard__CardWrapper-sc-77egjm-0",
                    "target": "div",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <div
                  className="StudyCard__CardWrapper-sc-77egjm-0 dzZgLX"
                >
                  <StudyCard__InputArea
                    name="question"
                    onChange={[Function]}
                    onKeyDown={[Function]}
                    placeholder="Question"
                    value=""
                  >
                    <StyledComponent
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "StudyCard__InputArea-sc-77egjm-1",
                            "isStatic": true,
                            "lastClassName": "fKGKTo",
                            "rules": Array [
                              "width:100%;border:none;background:transparent;color:#333;flex:1 auto;font-size:1em;font-family:",
                              "-apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"",
                              ";padding:0.5em 1em;resize:none;box-sizing:border-box;",
                            ],
                          },
                          "displayName": "StudyCard__InputArea",
                          "foldedComponentIds": Array [],
                          "render": [Function],
                          "styledComponentId": "StudyCard__InputArea-sc-77egjm-1",
                          "target": "textarea",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                      name="question"
                      onChange={[Function]}
                      onKeyDown={[Function]}
                      placeholder="Question"
                      value=""
                    >
                      <textarea
                        className="StudyCard__InputArea-sc-77egjm-1 fKGKTo"
                        name="question"
                        onChange={[Function]}
                        onKeyDown={[Function]}
                        placeholder="Question"
                        value=""
                      />
                    </StyledComponent>
                  </StudyCard__InputArea>
                  <StudyCard__InputArea
                    name="answer"
                    onChange={[Function]}
                    onKeyDown={[Function]}
                    placeholder="Answer"
                    value=""
                  >
                    <StyledComponent
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "StudyCard__InputArea-sc-77egjm-1",
                            "isStatic": true,
                            "lastClassName": "fKGKTo",
                            "rules": Array [
                              "width:100%;border:none;background:transparent;color:#333;flex:1 auto;font-size:1em;font-family:",
                              "-apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"",
                              ";padding:0.5em 1em;resize:none;box-sizing:border-box;",
                            ],
                          },
                          "displayName": "StudyCard__InputArea",
                          "foldedComponentIds": Array [],
                          "render": [Function],
                          "styledComponentId": "StudyCard__InputArea-sc-77egjm-1",
                          "target": "textarea",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                      name="answer"
                      onChange={[Function]}
                      onKeyDown={[Function]}
                      placeholder="Answer"
                      value=""
                    >
                      <textarea
                        className="StudyCard__InputArea-sc-77egjm-1 fKGKTo"
                        name="answer"
                        onChange={[Function]}
                        onKeyDown={[Function]}
                        placeholder="Answer"
                        value=""
                      />
                    </StyledComponent>
                  </StudyCard__InputArea>
                </div>
              </StyledComponent>
            </StudyCard__CardWrapper>
            <style__Button
              fullWidth={true}
              onClick={[Function]}
            >
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "style__Button-oz4l5n-5",
                      "isStatic": false,
                      "lastClassName": "gPTbBw",
                      "rules": Array [
                        "background:",
                        "#9e3bff",
                        ";border:none;border-radius:3px;padding:12px 15px;font-size:",
                        "0.8em",
                        ";font-weight:400;color:#fff;width:",
                        [Function],
                        ";&:hover{cursor:pointer;}",
                      ],
                    },
                    "displayName": "style__Button",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "style__Button-oz4l5n-5",
                    "target": "button",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
                fullWidth={true}
                onClick={[Function]}
              >
                <button
                  className="style__Button-oz4l5n-5 gPTbBw"
                  onClick={[Function]}
                >
                  Save
                </button>
              </StyledComponent>
            </style__Button>
          </section>
        </StyledComponent>
      </StudyCard__Wrapper>
    </StudyCard>
  </NotificationProvider>
</UserProvider>
`;
